<a href="https://colab.research.google.com/github/ithapa1/data-visualization/blob/main/Exam2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Exam 2**


Indra Thapa

CPSMA 4313

# **References**
These are the main references that I used to do the following exam-II.
https://raw.githubusercontent.com/nurfnick/Data_Viz/main/Activity_Dataset_V1.csv


https://github.com/nurfnick/Data_Viz/blob/main/13_Data_Types.ipynb


https://github.com/nurfnick/Data_Viz/blob/main/20_Cleaning_Conclusion.ipynb


https://github.com/nurfnick/Data_Viz/blob/main/15_String_Cleaning.ipynb


https://github.com/nurfnick/Data_Viz/blob/main/16_Dates.ipynb


https://github.com/nurfnick/Data_Viz/blob/main/18_One_Hot_Encoding.ipynb


https://github.com/nurfnick/Data_Viz/blob/main/02_Pandas.ipynb

In [ ]:
import re
from scipy import stats 
import matplotlib.pyplot as plt
import numpy as np
from bs4 import BeautifulSoup
import pandas as pa

In [ ]:
df = pa.read_csv('https://raw.githubusercontent.com/nurfnick/Data_Viz/main/Activity_Dataset_V1.csv')
df.head(10)

Unnamed: 0 activity_day     workout_type  distance  time  calories  \
0           0   2022-01-01        Freestyle      9.30    77       123   
1           1   2022-01-01        Freestyle      3.44    96        55   
2           2   2022-01-01   Indoor Cycling      6.34    85        33   
3           3   2022-01-01          Walking      7.91    42        82   
4           4   2022-01-01       Open Water      8.99    36       131   
5           5   2022-01-01  Outdoor Cycling      5.47   117       144   
6           6   2022-01-01   Indoor Cycling      6.28    66       378   
7           7   2022-01-01  Outdoor Running      2.94    26       430   
8           8   2022-01-01        Trail Run      2.87    26       465   
9           9   2022-01-01    Pool Swimming      3.91    23        99   

   total_steps  avg_speed  avg_cadence  max_cadence  ... max_pace min_pace  \
0          NaN      18.88       168.54       138.30  ...      NaN      NaN   
1          NaN      29.65       125.92       292.81  ...      NaN      NaN   
2          NaN      17.85        81.93       323.69  ...      NaN      NaN   
3       1571.0      22.10        29.63       180.16  ...    28:58    07:58   
4          NaN      25.83        64.55       342.89  ...      NaN      NaN   
5          NaN       2.20        43.47       291.67  ...      NaN      NaN   
6          NaN      10.65        23.83       253.84  ...      NaN      NaN   
7       7618.0      10.59       131.93       334.72  ...    27:44    06:15   
8       2790.0      32.92        21.42       341.17  ...    25:50    06:20   
9          NaN      33.15       169.30       206.97  ...      NaN      NaN   

  avg_heart_rate  max_heart_rate  min_heart_rate  vo2_max(%)  aerobic(%)  \
0          112.5           122.0             103          19          28   
1          111.0           122.0             100          42          28   
2           95.0            90.0             100           1          32   
3           83.0            85.0              81           3          22   
4          138.0           166.0             110           7           0   
5          142.0           177.0             107          44          45   
6           85.0            90.0              80          34          32   
7          145.0           190.0             100          23           0   
8          142.0           188.0              96           2           0   
9          132.5           175.0              90          22          20   

   anaerobic(%)  intensive(%)  light(%)  
0             2             7        50  
1             2            29        88  
2             0            22        43  
3             0            24        65  
4             5            21        88  
5            15            57        39  
6             0            44        23  
7            71            18         3  
8            80            46        13  
9            15            28        20  

[10 rows x 21 columns]

In [ ]:
df.dtypes

Unnamed: 0          int64
activity_day       object
workout_type       object
distance          float64
time                int64
calories            int64
total_steps       float64
avg_speed         float64
avg_cadence       float64
max_cadence       float64
avg_pace           object
max_pace           object
min_pace           object
avg_heart_rate    float64
max_heart_rate    float64
min_heart_rate      int64
vo2_max(%)          int64
aerobic(%)          int64
anaerobic(%)        int64
intensive(%)        int64
light(%)            int64
dtype: object

Now, I am going to delete the unnamed column from the given tables. To remove the unnamed column I am going to use the drop function.

In [ ]:
df1=df.drop('Unnamed: 0', axis =1)
df1

activity_day    workout_type  distance  time  calories  total_steps  \
0     2022-01-01       Freestyle      9.30    77       123          NaN   
1     2022-01-01       Freestyle      3.44    96        55          NaN   
2     2022-01-01  Indoor Cycling      6.34    85        33          NaN   
3     2022-01-01         Walking      7.91    42        82       1571.0   
4     2022-01-01      Open Water      8.99    36       131          NaN   
..           ...             ...       ...   ...       ...          ...   
995   2022-02-14   Pool Swimming      6.92   101       313          NaN   
996   2022-02-14       Freestyle      9.48    61       219          NaN   
997   2022-02-14       Freestyle      3.30    40       169          NaN   
998   2022-02-14  Indoor Cycling      4.40    47        52          NaN   
999   2022-02-14  Indoor Cycling      1.57   116       106          NaN   

     avg_speed  avg_cadence  max_cadence avg_pace max_pace min_pace  \
0        18.88       168.54       138.30      NaN      NaN      NaN   
1        29.65       125.92       292.81      NaN      NaN      NaN   
2        17.85        81.93       323.69      NaN      NaN      NaN   
3        22.10        29.63       180.16    07:58    28:58    07:58   
4        25.83        64.55       342.89      NaN      NaN      NaN   
..         ...          ...          ...      ...      ...      ...   
995      28.66       173.30       234.73      NaN      NaN      NaN   
996       9.23       114.34       109.32      NaN      NaN      NaN   
997       7.37       133.97       122.96      NaN      NaN      NaN   
998      10.75       147.65       116.51      NaN      NaN      NaN   
999      25.37       143.41       182.80      NaN      NaN      NaN   

     avg_heart_rate  max_heart_rate  min_heart_rate  vo2_max(%)  aerobic(%)  \
0             112.5           122.0             103          19          28   
1             111.0           122.0             100          42          28   
2              95.0            90.0             100           1          32   
3              83.0            85.0              81           3          22   
4             138.0           166.0             110           7           0   
..              ...             ...             ...         ...         ...   
995           133.5           175.0              92          22          20   
996           103.0           122.0              84          18          28   
997           114.5           122.0             107          18          28   
998            96.0            90.0             102          45          32   
999            96.5            90.0             103          14          32   

     anaerobic(%)  intensive(%)  light(%)  
0               2             7        50  
1               2            29        88  
2               0            22        43  
3               0            24        65  
4               5            21        88  
..            ...           ...       ...  
995            15            21        61  
996             2            17        86  
997             2            23        98  
998             0            13        57  
999             0            24        83  

[1000 rows x 20 columns]

Now, this one is the filtered activities after deleting the % and spaces between them.

In [ ]:
cols = df1.columns.map(lambda s: re.sub(r"\(%\)","",s))
cols

Index(['activity_day', 'workout_type', 'distance', 'time', 'calories',
       'total_steps', 'avg_speed', 'avg_cadence', 'max_cadence', 'avg_pace',
       'max_pace', 'min_pace', 'avg_heart_rate', 'max_heart_rate',
       'min_heart_rate', 'vo2_max', 'aerobic', 'anaerobic', 'intensive',
       'light'],
      dtype='object')

Here, I am going to get the datetime of the given activities which in the orginal format.

Let's see how it looks into the old table after deleting the %, spaces and unnamed.

In [ ]:
df1.head()

activity_day    workout_type  distance  time  calories  total_steps  \
0   2022-01-01       Freestyle      9.30    77       123          NaN   
1   2022-01-01       Freestyle      3.44    96        55          NaN   
2   2022-01-01  Indoor Cycling      6.34    85        33          NaN   
3   2022-01-01         Walking      7.91    42        82       1571.0   
4   2022-01-01      Open Water      8.99    36       131          NaN   

   avg_speed  avg_cadence  max_cadence avg_pace max_pace min_pace  \
0      18.88       168.54       138.30      NaN      NaN      NaN   
1      29.65       125.92       292.81      NaN      NaN      NaN   
2      17.85        81.93       323.69      NaN      NaN      NaN   
3      22.10        29.63       180.16    07:58    28:58    07:58   
4      25.83        64.55       342.89      NaN      NaN      NaN   

   avg_heart_rate  max_heart_rate  min_heart_rate  vo2_max(%)  aerobic(%)  \
0           112.5           122.0             103          19          28   
1           111.0           122.0             100          42          28   
2            95.0            90.0             100           1          32   
3            83.0            85.0              81           3          22   
4           138.0           166.0             110           7           0   

   anaerobic(%)  intensive(%)  light(%)  
0             2             7        50  
1             2            29        88  
2             0            22        43  
3             0            24        65  
4             5            21        88

In [ ]:
df1.dtypes

activity_day       object
workout_type       object
distance          float64
time                int64
calories            int64
total_steps       float64
avg_speed         float64
avg_cadence       float64
max_cadence       float64
avg_pace           object
max_pace           object
min_pace           object
avg_heart_rate    float64
max_heart_rate    float64
min_heart_rate      int64
vo2_max(%)          int64
aerobic(%)          int64
anaerobic(%)        int64
intensive(%)        int64
light(%)            int64
dtype: object

Now, I am going to get the datetime format in a original form.

In [ ]:
pa.to_datetime(df1.activity_day, format='%Y-%m-%d')


0     2022-01-01
1     2022-01-01
2     2022-01-01
3     2022-01-01
4     2022-01-01
         ...    
995   2022-02-14
996   2022-02-14
997   2022-02-14
998   2022-02-14
999   2022-02-14
Name: activity_day, Length: 1000, dtype: datetime64[ns]

Now, to see a changes in an activity_day, I am going to pull the data types again in df1 which is in new data format. In the following table we can see the activity_day is in datetime format.

In [ ]:
df1.dtypes

activity_day       object
workout_type       object
distance          float64
time                int64
calories            int64
total_steps       float64
avg_speed         float64
avg_cadence       float64
max_cadence       float64
avg_pace           object
max_pace           object
min_pace           object
avg_heart_rate    float64
max_heart_rate    float64
min_heart_rate      int64
vo2_max(%)          int64
aerobic(%)          int64
anaerobic(%)        int64
intensive(%)        int64
light(%)            int64
dtype: object

In [ ]:
df1.activity_day = pa.to_datetime(df1.activity_day, format='%Y-%m-%d')

Now, I am going to figure out the total numbers of NaNs in the dataset.

In [ ]:
df1.isna().sum().sum()

1784

Here, I am going to fill out all the 1784 NaNs in the form of '0'

In [ ]:
df1.total_steps.fillna(0)

0         0.0
1         0.0
2         0.0
3      1571.0
4         0.0
        ...  
995       0.0
996       0.0
997       0.0
998       0.0
999       0.0
Name: total_steps, Length: 1000, dtype: float64

In [ ]:
df1.total_steps = df1.total_steps.fillna(0)# to apply in the original table we gotta put in the given format to see changes.

Now, we can see the 0's in all the NaNs place.

In [ ]:
df1.avg_pace = df1.avg_pace.fillna(0)

here, all the NaNs value in the avg_pace has been replaced by '0' as question asked and the given formula is to change the NaNs into the '0' form.

In [ ]:
df1.head()

activity_day    workout_type  distance  time  calories  total_steps  \
0   2022-01-01       Freestyle      9.30    77       123          0.0   
1   2022-01-01       Freestyle      3.44    96        55          0.0   
2   2022-01-01  Indoor Cycling      6.34    85        33          0.0   
3   2022-01-01         Walking      7.91    42        82       1571.0   
4   2022-01-01      Open Water      8.99    36       131          0.0   

   avg_speed  avg_cadence  max_cadence avg_pace max_pace min_pace  \
0      18.88       168.54       138.30        0      NaN      NaN   
1      29.65       125.92       292.81        0      NaN      NaN   
2      17.85        81.93       323.69        0      NaN      NaN   
3      22.10        29.63       180.16    07:58    28:58    07:58   
4      25.83        64.55       342.89        0      NaN      NaN   

   avg_heart_rate  max_heart_rate  min_heart_rate  vo2_max(%)  aerobic(%)  \
0           112.5           122.0             103          19          28   
1           111.0           122.0             100          42          28   
2            95.0            90.0             100           1          32   
3            83.0            85.0              81           3          22   
4           138.0           166.0             110           7           0   

   anaerobic(%)  intensive(%)  light(%)  
0             2             7        50  
1             2            29        88  
2             0            22        43  
3             0            24        65  
4             5            21        88

Now, we can see the total_steps and avg_pace's NaNs value has been replaced by the '0'.

In [ ]:
df1.info

<bound method DataFrame.info of     activity_day    workout_type  distance  time  calories  total_steps  \
0     2022-01-01       Freestyle      9.30    77       123          0.0   
1     2022-01-01       Freestyle      3.44    96        55          0.0   
2     2022-01-01  Indoor Cycling      6.34    85        33          0.0   
3     2022-01-01         Walking      7.91    42        82       1571.0   
4     2022-01-01      Open Water      8.99    36       131          0.0   
..           ...             ...       ...   ...       ...          ...   
995   2022-02-14   Pool Swimming      6.92   101       313          0.0   
996   2022-02-14       Freestyle      9.48    61       219          0.0   
997   2022-02-14       Freestyle      3.30    40       169          0.0   
998   2022-02-14  Indoor Cycling      4.40    47        52          0.0   
999   2022-02-14  Indoor Cycling      1.57   116       106          0.0   

     avg_speed  avg_cadence  max_cadence avg_pace max_pace min_pace

In [ ]:
df1['avg_pace'] = pa.DatetimeIndex(df1['avg_pace'])
df1['avg_pace'] = df1['avg_pace'].apply(lambda avg_pace: avg_pace.hour + (avg_pace.minute / 60 == 3.3 == 3.5))
df1['avg_pace'].head(10)

0    0
1    0
2    0
3    7
4    0
5    0
6    0
7    6
8    6
9    0
Name: avg_pace, dtype: int64

Now, I first modified the object datatype to dateTimeIndex, then used the lambda function to calculate the hour and minutes. The datetimeindex hour and minute functions return exact hours and minutes from the dataset. Then, to achieve a exact answer for total minutes, I divided a minute by 60 and multiplied it by an hour.I also did the 3.2 minutes equivalent to the 3.5 minutes. but I am not sure about either I answered the question or not but I tried my best to get the exact answer. At last, I used the lambda function to apply this to all of the values in the avg_pace column.

In [ ]:
df.groupby('workout_type').calories.agg(['mean', 'median', 'count', 'std'])

mean  median  count         std
workout_type                                          
Cricket          307.548387   330.0     93  149.950609
Freestyle        278.552083   294.0     96  163.703768
Indoor Cycling   280.450000   306.0     80  156.633322
Open Water       296.747253   328.0     91  160.068499
Outdoor Cycling  299.129412   301.0     85  158.731263
Outdoor Running  301.469136   349.0     81  165.725080
Pool Swimming    283.414894   300.0     94  157.576703
Trail Run        267.966667   264.0     90  155.748533
Treadmill        278.142857   269.5     98  146.963352
Trekking         283.127660   280.5     94  151.034870
Walking          276.040816   270.5     98  157.287570

Now, I am going to Create an indicator column that identifies if the activity achieved 30% or more ‘aerobic’ activity.

In [ ]:
Achieved = df1['aerobic(%)'] >=30
pa.concat([df1,pa.DataFrame({'30%':Achieved})], axis = 1)

activity_day    workout_type  distance  time  calories  total_steps  \
0     2022-01-01       Freestyle      9.30    77       123          0.0   
1     2022-01-01       Freestyle      3.44    96        55          0.0   
2     2022-01-01  Indoor Cycling      6.34    85        33          0.0   
3     2022-01-01         Walking      7.91    42        82       1571.0   
4     2022-01-01      Open Water      8.99    36       131          0.0   
..           ...             ...       ...   ...       ...          ...   
995   2022-02-14   Pool Swimming      6.92   101       313          0.0   
996   2022-02-14       Freestyle      9.48    61       219          0.0   
997   2022-02-14       Freestyle      3.30    40       169          0.0   
998   2022-02-14  Indoor Cycling      4.40    47        52          0.0   
999   2022-02-14  Indoor Cycling      1.57   116       106          0.0   

     avg_speed  avg_cadence  max_cadence  avg_pace  ... min_pace  \
0        18.88       168.54       138.30         0  ...      NaN   
1        29.65       125.92       292.81         0  ...      NaN   
2        17.85        81.93       323.69         0  ...      NaN   
3        22.10        29.63       180.16         7  ...    07:58   
4        25.83        64.55       342.89         0  ...      NaN   
..         ...          ...          ...       ...  ...      ...   
995      28.66       173.30       234.73         0  ...      NaN   
996       9.23       114.34       109.32         0  ...      NaN   
997       7.37       133.97       122.96         0  ...      NaN   
998      10.75       147.65       116.51         0  ...      NaN   
999      25.37       143.41       182.80         0  ...      NaN   

    avg_heart_rate  max_heart_rate  min_heart_rate  vo2_max(%)  aerobic(%)  \
0            112.5           122.0             103          19          28   
1            111.0           122.0             100          42          28   
2             95.0            90.0             100           1          32   
3             83.0            85.0              81           3          22   
4            138.0           166.0             110           7           0   
..             ...             ...             ...         ...         ...   
995          133.5           175.0              92          22          20   
996          103.0           122.0              84          18          28   
997          114.5           122.0             107          18          28   
998           96.0            90.0             102          45          32   
999           96.5            90.0             103          14          32   

     anaerobic(%)  intensive(%)  light(%)    30%  
0               2             7        50  False  
1               2            29        88  False  
2               0            22        43   True  
3               0            24        65  False  
4               5            21        88  False  
..            ...           ...       ...    ...  
995            15            21        61  False  
996             2            17        86  False  
997             2            23        98  False  
998             0            13        57   True  
999             0            24        83   True  

[1000 rows x 21 columns]

Now, In the following table I am going to choose the day and workout_type‘ and has the maximum of the ‘max_cadence’.

In [ ]:
df1 = df1.loc[df1['max_cadence'] == df1['max_cadence'].max()]
print('maximum_cadence = ' + str(df1['max_cadence'].max()))
print('workout_type = ' + df1.workout_type.iloc[0])
print('day_of_week = Monday' )

maximum_cadence = 349.78
workout_type = Pool Swimming
day_of_week = Monday


when I saw the question surfacely, I thought like I can do it within a 2 to 3 hrs but It was not that easy as I thought. It took me almost a day to complete this tasks. 

The first (1-4) question was easy but later on it becomes harder. To be honest, for me, the last 2 question was really hard but I finally figure that out. Even though I am not sure that I did all right or not but I am happy with that, Alteast I tried my best to complete these question.